### PR TITLE
COMPAT: remove Categorical pickle compat with Pandas < 0.16

### PR DESCRIPTION
**Pull request by topper-123**
- [ ] closes #xxxx
- [ ] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Removes Categorical pickle compat with Pandas < 0.16.

I've added a ``categorical.pickle`` file to replace the deprecated file. It was create like this:

```python
>>> cat = pd.Categorical(['a', 'b', 'c', 'd'])
>>> pickle.dump(cat, open(<filepath>, 'wb'))
```

### CODE REVIEW

**Comment by jbrockmendel** [comment]
why close?  ive been looking forward to seeing these big chunks of pickle compat code removed

**Comment by topper-123** [comment]
Sorry, I saw there were errors, and don't have time today to figure them out and didn't want to leave a failing PR open too long.

I'll open the PR again, when I figure out the problem.

**Comment by topper-123** [comment]
Reopened as a draft. Please hold review until I get the bug fixed.

**Comment by topper-123** [comment]
Reopened and tests passed.

**Comment by topper-123** [comment]
@jbrockmendel , agree with these changes?

File: pandas/tests/io/test_common.py [hunk 1]
```diff
@@ -222,7 +222,7 @@ def test_read_expands_user_home_dir(
             (pd.read_sas, "os", ("io", "sas", "data", "test1.sas7bdat")),
             (pd.read_json, "os", ("io", "json", "data", "tsframe_v012.json")),
             (pd.read_msgpack, "os", ("io", "msgpack", "data", "frame.mp")),
-            (pd.read_pickle, "os", ("io", "data", "categorical_0_14_1.pickle")),
+            (pd.read_pickle, "os", ("io", "data", "categorical.pickle")),
```
Comment (jbrockmendel) [inline pandas/tests/io/test_common.py]: should there be a version number on this?

**Comment by jbrockmendel** [comment]
Small question, otherwise LGTM

File: pandas/tests/io/test_common.py (hunk 1 above)
Comment (topper-123) [reply pandas/tests/io/test_common.py]: Yeah, that will  probably be easier for future maintenance. Will add it.

**Comment by topper-123** [comment]
I’ve renamed the file.

**Comment by topper-123** [comment]
@jbrockmendel ? Ok to pull in?

**Comment by jbrockmendel** [comment]
pickle compat is way outside my area of expertise.  @jorisvandenbossche?

File: pandas/core/arrays/categorical.py [hunk 2]
```diff
@@ -1350,24 +1350,7 @@ def __setstate__(self, state):
...
-        # 0.16.0 ordered change
-        if "_ordered" not in state:
-
-            # >=15.0 < 0.16.0
-            if "ordered" in state:
-                state["_ordered"] = state.pop("ordered")
-            else:
-                state["_ordered"] = False
-
-        # 0.21.0 CategoricalDtype change
```
Comment (jreback) [inline pandas/core/arrays/categorical.py]: why is there a 025 pickle
file included? this is covered by the legacy pickle dukes no?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/categorical.py (hunk 2 above)
Comment (topper-123) [reply pandas/core/arrays/categorical.py]: The test ``test_read_fspath_all`` tests if reading custom file paths (``CustomFS``) returns an expected value. Seems reasonable that a pickle are included for that testing. Adding 0.25 in the name is not strictly necessary, but can possibly be handy info to have.

Probably this test could instead test if temp files are read correctly back in, but that would be for another PR, IMO.

Comment (jreback) [reply pandas/core/arrays/categorical.py]: we already have round trip categorical pickle tests

Comment (topper-123) [reply pandas/core/arrays/categorical.py]: Didn't mean test round-tripping, but use roundtripping in this test, rather than keeping a file in /data.

**Review by WillAyd — APPROVED** [review]
lgtm. I don't have a problem with the inclusion of the 0.25 file but am indifferent if we want to do away with it as well

File: doc/source/whatsnew/v1.0.0.rst [hunk 3]
```diff
@@ -67,6 +67,8 @@ Removal of prior version deprecations/changes
 - :meth:`pandas.Series.str.cat` does not accept list-likes *within* list-likes anymore (:issue:`27611`)
 - Removed the previously deprecated :meth:`ExtensionArray._formatting_values`. Use :attr:`ExtensionArray._formatter` instead. (:issue:`23601`)
 - Removed the previously deprecated ``IntervalIndex.from_intervals`` in favor of the :class:`IntervalIndex` constructor (:issue:`19263`)
+- Ability to read pickles containing :class:`Categorical` instances created with pre-0.16 version of pandas has been removed (:issue:`27538`)
```
[the remark this answers is not in the document]
Comment (topper-123) [reply doc/source/whatsnew/v1.0.0.rst]: IMO it is ok to include this, given that something won't work that previouly did worked...

**Comment by WillAyd** [comment]
@topper-123 can you merge master on this to be safe? If no other feedback in 2 days let's get this merged

**Comment by topper-123** [comment]
I've rebased.

**Comment by WillAyd** [comment]
Thanks @topper-123

**Comment by jbrockmendel** [comment]
There are other <0.16 pickle tests out there.  Can we get rid of those too?  `tests.tseries.test_offsets.test_pickle_v0_15_2` is a blocker for cythonizing DateOffset.

**Comment by WillAyd** [comment]
I think so - we are only guaranteed backwards compatible to pandas 0.20.3 as of the 0.25 release

see #27082

**Pull request merged**